core/trino-main/src/main/java/io/trino/sql/planner/EqualityInference.java

```diff
@@ -53,25 +53,21 @@
 public class EqualityInference
 {
     // Comparator used to determine Expression preference when determining canonicals
-    private static final Comparator<Expression> CANONICAL_COMPARATOR = Comparator
-            // Current cost heuristic:
-            // 1) Prefer fewer input symbols
-            // 2) Prefer smaller expression trees
-            // 3) Sort the expressions alphabetically - creates a stable consistent ordering (extremely useful for unit testing)
-            // TODO: be more precise in determining the cost of an expression
-            .comparingInt((ToIntFunction<Expression>) (expression -> SymbolsExtractor.extractAll(expression).size()))
-            .thenComparingLong(expression -> SubExpressionExtractor.extract(expression).count())
-            .thenComparing(Expression::toString);
-
+    private final Comparator<Expression> canonicalComparator;
     private final Multimap<Expression, Expression> equalitySets; // Indexed by canonical expression
     private final Map<Expression, Expression> canonicalMap; // Map each known expression to canonical expression
     private final Set<Expression> derivedExpressions;
+    private final CachingSubExpressionExtractor subExpressionExtractor;
+    private final CachingSymbolsExtractor symbolsExtractor;
 
-    private EqualityInference(Multimap<Expression, Expression> equalitySets, Map<Expression, Expression> canonicalMap, Set<Expression> derivedExpressions)
+    private EqualityInference(Multimap<Expression, Expression> equalitySets, Map<Expression, Expression> canonicalMap, Set<Expression> derivedExpressions, Comparator<Expression> canonicalComparator, CachingSubExpressionExtractor subExpressionExtractor, CachingSymbolsExtractor symbolsExtractor)
     {
         this.equalitySets = equalitySets;
         this.canonicalMap = canonicalMap;
         this.derivedExpressions = derivedExpressions;
+        this.canonicalComparator = canonicalComparator;
+        this.subExpressionExtractor = subExpressionExtractor;
+        this.symbolsExtractor = symbolsExtractor;
```

Comment:
As far as I can tell, these don't depend on any external state. So, instead of building them and passing them in from the `newEqualityInference()` method, just create them here.

Reply:
It does not depend on external state. But, `symbolsCache` and `subExpressionCache` is used as well by `canonicalComparator`that is created in `newInstance` method. `canonicalComparator` could be created in constructor as well, but I've put this into kind of fabric method. If you really want to move a bit more "complex"   stuff from fabric method to constructor, let me sign, I will do it.